### PR TITLE
Usage statistics plugin

### LINKED ISSUE #75: GCS: Analytics
Consider instrumenting GCS with (Google?) analytics. This would allow us to track usage of particular boards, features, and errors, allowing us to make more informed decisions and better data to combat frustrations.

### PR DESCRIPTION
**Pull request by PTDreamer**
[Review on Reviewable](https://reviewable.io/reviews/d-ronin/dronin/339)

edit by mlyle: fixes #75

### CODE REVIEW

**Comment by dustin** [comment]
Review status: 0 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 237 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6HN3g4LoQieRgAUURw:-K6HN3g4LoQieRgAUURx:-66663676) ([raw file](https://github.com/d-ronin/dronin/blob/94ef5e456fdc15329c4e616ef45d1645dcbb4361/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L237)):*
I don't think we need any IP info.  The client will contact us from an address.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 254 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6HO2YxeOQt_JJuGzzi:-K6HO2YxeOQt_JJuGzzj:-1926476527) ([raw file](https://github.com/d-ronin/dronin/blob/94ef5e456fdc15329c4e616ef45d1645dcbb4361/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L254)):*
You can just omit information that's shouldn't be sent.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
This looks super nice and is everything we could hope for, I think.  I agree with @dustin's review comments.  This one needs some thorough test to ensure resource usage, etc, is not a problem.   I'll be running it in the coming days.  (Everyone, don't laugh at me for all my spurious/repetitive/silly UI actions :)

**Comment by mlyle** [comment]
```
linking ../../../lib/dr/plugins/TauLabs/libUsageStats.so
/usr/bin/ld: cannot find -lUAVObjects
/usr/bin/ld: cannot find -lUAVTalk
/usr/bin/ld: cannot find -lUAVObjectUtil
/usr/bin/ld: cannot find -lUploader
collect2: error: ld returned 1 exit status
make[4]: *** [../../../lib/dr/plugins/TauLabs/libUsageStats.so] Error 1
```

**Comment by dustin** [comment]
The server side is half-done.  The storage is all async, so as long as valid JSON is posted, you should get an HTTP 202 response and it'll be queued for further processing.  Feel free to blast up a sample.

**Comment by PTDreamer** [comment]
> The server side is half-done. The storage is all async, so as long as valid JSON is posted, you should get an HTTP 202 response and it'll be queued for further processing. Feel free to blast up a sample.

Sent one, is there a way to have a look at the data received?

File: ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+        b["time"] = board.time.toString();
+        b["name"] = board.board.board.data()->getBoardNameFromID(board.device.boardID());
+        b["uavoHash"] = QString(board.device.uavoHash.toHex());
+        b["ID"] = board.device.boardID();
+        b["fwHash"] = QString(board.device.fwHash.toHex());
+        b["gitDate"] = board.device.gitDate;
+        b["gitHash"] = board.device.gitHash;
+        b["gitTag"] = board.device.gitTag;
+        if(sendPrivateData)
+            b["CPU"] = board.board.cpu_serial;
```
Comment (mluessi) [inline ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp]: Can we send a hash (e.g. SHA-512) of the serial number instead? In the future, I plan to offer the possibility of people registering their brainfpv flight controllers and the serial number will be required for this, so making the serial public may allow people to register using the serial number from someone else. Using a hash would give the same functionality, i.e., we could count the number of unique boards etc.

**Comment by mluessi** [comment]
Also, I think sending usage statistics should be disabled by default, or there should be a dialog box when the GCS is first started asking the user if it is OK to send the data (it could default to yes, so clicking OK would enough). We should be careful to to appear creepy..

**Comment by dustin** [comment]
Dialog box is OK, but I suspect we'll get a lot more useful info if it's on
by default and they're asked to turn it off than the other way around.  One
way is asking them to be spied on, while the other is asking them if
they're paranoid.

We're still going to get usage information when they call out for news. We
should put with stuff in that request to provide possibly more custom news
feeds while noting what's out there.

For this, I don't know that pii is really that interesting.  Knowing
hardware and modules and OSes is good.  I'm going to know an IP address
which tells us roughly where they are (sometimes city level).  What do we
get by tracking an individual board?

**Comment by mlyle** [comment]
What waffle does with tags is annoying and we need to figure it out.

**Comment by PTDreamer** [comment]
Translation?

**Comment by mlyle** [comment]
@PTDreamer -- not your fault, it just likes putting "ready" tags and "in progress" tags on things that don't really make sense.

But it makes this pretty chart of what's happening so I forgive it: https://waffle.io/d-ronin/dRonin?milestone=Renatus

**Comment by dustin** [comment]
I saw this earlier:

```
{
    "IP": "95.93.153.125",
    "boardsSeen": [
    ],
    "gcs_revision": "usage_stats%@%20151224-110f0366",
    "gcs_version": "usage_stats:110f0366 20151224",
    "widgets": [
        {
            "className": "QTabBar",
            "index": "1",
            "objectName": "qt_tabwidget_tabbar",
            "parentName": "",
            "text": "Flight data",
            "time": "Thu Dec 24 13:09:12 2015"
        },
        {
            "className": "QTabBar",
            "index": "2",
            "objectName": "qt_tabwidget_tabbar",
            "parentName": "",
            "text": "Configuration",
            "time": "Thu Dec 24 13:09:12 2015"
        },
        {
            "className": "QTabBar",
            "index": "3",
            "objectName": "qt_tabwidget_tabbar",
            "parentName": "",
            "text": "System",
            "time": "Thu Dec 24 13:09:13 2015"
        },
        {
            "className": "QTabBar",
            "index": "4",
            "objectName": "qt_tabwidget_tabbar",
            "parentName": "",
            "text": "Scopes",
            "time": "Thu Dec 24 13:09:15 2015"
        },
        {
            "className": "QTabBar",
            "index": "5",
            "objectName": "qt_tabwidget_tabbar",
            "parentName": "",
            "text": "Advanced",
            "time": "Thu Dec 24 13:09:16 2015"
        }
    ]
}
```

But more recently:

```
{
    "IP": "95.93.153.125",
    "boardsSeen": [
    ],
    "gcs_revision": "usage_stats%@%20151224-110f0366",
    "gcs_version": "usage_stats:110f0366 20151224",
    "widgets": [
    ]
}
```

I think we want OS in there.  Not sure what else.  IP probably isn't very useful, as it might often be a 10.x or similar.

**Comment by PTDreamer** [comment]
The first log is just starting and clicking on the workspace tabs, the second is just starting and closing the gcs.
The IP is the external IP so it will never be 10.x or 192.x
I'll add the OS info, the same I added to the crash reports

**Comment by PTDreamer** [comment]
@mlyle that is a nice tool, another thing for my browser favorites

**Comment by dustin** [comment]
Review status: 15 of 16 files reviewed at latest revision, 4 unresolved discussions.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 243 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6Lj0roUdqh-YPwEVq4:-K6Lj0roUdqh-YPwEVq5:1772557996) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L243)):*
Can we keep time in a consistent format?  RFC3339 in UTC with the timezone marker is a bit more useful, as we don't have timezone and stuff here.

We'll probably want a toplevel time as well since we're not sure how far skewed the clock is.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by PTDreamer** [comment]
Review status: 15 of 16 files reviewed at latest revision, 4 unresolved discussions.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 237 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6HN3g4LoQieRgAUURw:-K6O0wTrZIEaU5NBerCg:324105205) ([raw file](https://github.com/d-ronin/dronin/blob/94ef5e456fdc15329c4e616ef45d1645dcbb4361/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L237)):*
On the plugin options the user is given the choice to share private data (IP and CPU ID) so even though the server will always know the client's IP it still needs a way to know if the user wants to allow it's storage and usage. This is one of the ways to do it although I can just send a bool.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 254 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6HO2YxeOQt_JJuGzzi:-K6O3IrABTDpo0W9uXMa:160369988) ([raw file](https://github.com/d-ronin/dronin/blob/94ef5e456fdc15329c4e616ef45d1645dcbb4361/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L254)):*
ACK

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 243 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6Lj0roUdqh-YPwEVq4:-K6O1o8QHdZ1vIsCz8Lr:-1386813577) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L243)):*
That is an easy change although I don't see the point. The time will be used to sort log events in relation to each other so timezone is irrelevant. Do you see a reason to be able to correlate logs sent from different users?   

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 252 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6G-I8LSIFOTxqXJ7k7-r2-252:-K6O2vjnMF8soOoeMpXn:1586062786) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L252)):*
We can send whatever. What do others think?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
Build fails on OSX in default configuration:

```
make[4]: *** No rule to make target `/Users/mlyle/dronin/ground/gcs/src/plugins/usagestatsgadget/UsageStats_debug.pluginspec', needed by `../../../bin/dRonin-GCS.app/Contents/Plugins/TauLabs/UsageStats_debug.pluginspec'.  Stop.
```

I looked at it for a couple of minutes but couldn't figure out why.

**Comment by mlyle** [comment]
In the release configuration, it seems to work fine.  @dustin, is there a way I can look at the data?

**Comment by mlyle** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 237 \[r1\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6HN3g4LoQieRgAUURw:-K6Rq5bOdXNkTP3ojnOU:-978838417) ([raw file](https://github.com/d-ronin/dronin/blob/94ef5e456fdc15329c4e616ef45d1645dcbb4361/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L237)):*
I think I prefer the bool, just because it eliminates a dependency on an external server and another round trip.  This way there'll only be one web request that can fail or hang.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 243 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6Lj0roUdqh-YPwEVq4:-K6RpkIejOSs383YmVbx:72639178) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L243)):*
Could be different machines for the same user-- I think that's a pretty common use case.  Also whether it's before/after we announce something is interesting.  It's not essential, but it would be nice if we could normalize times.

---

*[ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp, line 252 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6G-I8LSIFOTxqXJ7k7-r2-252:-K6Rpy7c-oLC71eU2RE2:-1439654191) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/usagestatsgadget/usagestatsplugin.cpp#L252)):*
I would prefer to send the serial number and do any anonymizing on the server side.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
*[ground/gcs/src/plugins/plugins.pro, line 345 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6RqTWREt-t-Av24UYb:-K6RqTWREt-t-Av24UYc:1589832624) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/plugins.pro#L345)):*
Just noticed-- this should be up above the board plugins.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
This is super exciting and close-- I hope to be able to merge it soon.

Tested cases with good, bad, and hanging network connectivity and it seemed OK on OS X.  Just need to see the actual usage logs and get things cleaned up from review comments.

This is great stuff that should really help us improve the GCS @PTDreamer --- way more than I hoped for  in our first release.

**Comment by dustin** [comment]
@mlyle were you able to see the entries in the queue in the admin console?

**Comment by mlyle** [comment]
Ah, I didn't realize I needed to look at the GAE task queue.  (Had been looking at datastore).  Yes, they're there and nice.

**Comment by dustin** [comment]
Yeah, sorry that want clear.  I wanted to get some samples before figuring
out how to store them since I don't think we'll have columns for most
things.  I'll just extract a couple we expect to use for querying and store
the rest raw.

**Comment by mlyle** [comment]
I think there's some things that would be nice to roll up automatically-- number of users, number of boards, what percentage of runs a control is touched in, what percentage of users have touched a control ever.

**Comment by PTDreamer** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[ground/gcs/src/plugins/plugins.pro, line 345 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6RqTWREt-t-Av24UYb:-K6ViTrRu1_K0MkK-OZ9:1502162139) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/plugins.pro#L345)):*
Historically new plugins are added to the end of the list. Since the template isn't "ordered" it really doesn't matter. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by PTDreamer** [comment]
@mlyle please confirm the build error occurs on even a clean build. That file shouldn't be treated as a source file, it is on the .pro under "OTHER_FILES" which only makes it show under QtCreator project explorer. 

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339#-:-K6Vpq2F05Qg8uvBQIoR:1438040317)*

**Comment by mlyle** [comment]
Review status: 14 of 16 files reviewed at latest revision, 4 unresolved discussions.

---

*[ground/gcs/src/plugins/plugins.pro, line 345 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6RqTWREt-t-Av24UYb:-K6XNu2-L1bmap_QqnMk:292074935) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/plugins.pro#L345)):*
It just seems nice to keep all the board plugins together and have the above section header for it.  If you don't really want to change it (or just want to remove the board plugins section header) that's fine.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
It happens with a clean build-- both of the PR branch and of the PR branch merged to next.

I don't really know why it's happening just in this plugin.

**Comment by PTDreamer** [comment]
Should be fixed now

---

Review status: 11 of 16 files reviewed at latest revision, 5 unresolved discussions.

---

*[ground/gcs/src/plugins/plugins.pro, line 345 \[r2\]](https://reviewable.io:443/reviews/d-ronin/dronin/339#-K6RqTWREt-t-Av24UYb:-K6jw86EgzcyI4IQ-bhK:-490826578) ([raw file](https://github.com/d-ronin/dronin/blob/110f036693aa90547d145a2ca8fe8062a6109af7/ground/gcs/src/plugins/plugins.pro#L345)):*
Moved

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339#-:-K6jw4LtDHVw3GTUYTIQ:-1056516971)*

**Comment by PTDreamer** [comment]
Any issues remaining?

**Comment by mlyle** [comment]
This looks good, but I want to review the diff one more time and to try it again just in case.

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
Reviewed 1 of 2 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339)*

**Comment by mlyle** [comment]
:lgtm_strong: 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/339#-:-K6k4MXsgRmv9iLpzSZi:-1868698381)*

**Comment by mlyle** [comment]
Tested, seems to work very nicely.  Thanks, @PTDreamer --- this will be very helpful in figuring out how the GCS is used.

**Pull request merged**